packages/medusa/src/services/cart.ts

```diff
@@ -1739,77 +1754,120 @@ class CartService extends TransactionBaseService {
         )
 
         const { total, region } = cart
+
+        const providerSet = new Set(region.payment_providers.map((p) => p.id))
+
+        /**
+         * Find the payment sessions that needs to be removed
+         *
+         * if the cart total is <= 0 or the provider does not belong to the region then delete the sessions.
+         * The deletion occurs locally if there is no external data or if it is not selected
+         * otherwise the deletion will also occur remotely through the external provider.
+         */
+
+        const paymentSessionsToRemove = cart.payment_sessions.filter(
+          (session) => {
+            return total <= 0 || !providerSet.has(session.provider_id)
```

Comment:
question: If I am not mistaken, if the total is below 0, all payment sessions are removed (as this would always return true), so would it make sense to extract this into a preliminary check?

Reply:
Already done, i need to push 👌

Reply:
in fact I ve pushed already, I gather the sessions that need to be removed. After they are removed I return early.

Reply:
l 1788

Reply:
Wouldn't we be able to do: 
```ts
if (total <= 0) {
    await Promise.all(
          paymentSessionsToRemove.map(async (session) => {
            if (session.is_selected || Object.keys(session.data).length) {
              return this.paymentProviderService_
                .withTransaction(transactionManager)
                .deleteSession(session)
            }

            return psRepo.delete(session)
          })
        )
}
```

To save having to loop through the sessions twice?

Reply:
I ve cleaned up a bit